### PR TITLE
4 Send Questions To The Participant

### PR DESCRIPTION
**Pull request by leonardk1**
#4 
As an Administrator
I Want To Send Participants Questions
  When They Have Finished Answering The Previous Questions
So That I can Collect all the Data needed for the campaign for the participants

##### Acceptance Criteria
- [x] GIVEN I Send Participants questions, THEN The Following Should be put into consideration
  - if the Participant Has Already received the first question, then they should not be resent the first question if they have not taken part in the campaign as yet
 - the participant should only get the next question if they answer the previous
 - When the participants has finished answering the questions, they should not be resent the questions

### CODE REVIEW

**Comment by leonardk1** [comment]
@godfreym  please review, This is the Heart of this application, so you might check out the whole code, closely :wink: ,
All yours

File: app/controllers/messages_controller.rb [hunk 1]
```diff
@@ -3,44 +3,41 @@ class MessagesController < ApplicationController
...
-    @options = @question_one.options
-    @option_string = ""
-    @counter = 1
-    @options.each do |option|
-      @option_string << "\n #{@counter}. #{option.title}"
-      @counter += 1
-    end
-
-    @msg = "#{@question_one.title} \n #{@option_string}"
     @participants = Participant.all
```
Comment (godfreym) [inline app/controllers/messages_controller.rb]: @leonardk1 Are all participants stored in the table before they can participate on a campaign?

Comment (godfreym) [inline app/controllers/messages_controller.rb]: @leonardk1 can some one participate on a campaign if he/she is not a stored a member ?

Comment (leonardk1) [inline app/controllers/messages_controller.rb]: @godfreym  for the first question, yes they are all stored, and for the second question for one to participate, the administrator has to add them to the participant table
and we use yml for this

File: app/controllers/messages_controller.rb [hunk 2]
```diff
@@ -3,44 +3,41 @@ class MessagesController < ApplicationController
...
   # Process incoming SMS
-    def process_message(participant)
-      @campaign = Campaign.find(participant.current_campaign)
-      if participant.question_count < @campaign.questions.count
-        @new_question_count = participant.question_count + 1
-        @message = @campaign.questions[@new_question_count].title
-        respond(@message)
-        participant.track_campaign(@campaign.id, @new_question_count)
-      else
+    def process_message(participant, answer)
```
Comment (godfreym) [inline app/controllers/messages_controller.rb]: @leonardk1 what does `process_message` do in terms of customer voice?

File: app/models/participant.rb [hunk 3]
```diff
@@ -15,4 +15,60 @@ def send_message(msg)
...
+  def last_answer_entry
+    return Answer.where(participant_id: self.id).order("created_at").last
+  end
+
+  def tracked_campaign
+    return Campaign.find(self.current_campaign)
+  end
+
+  def campaign_incomplete?
+    @number_of_questions = self.tracked_campaign.questions.count
```
Comment (godfreym) [inline app/models/participant.rb]: @leonardk1 could you please explain this whole condition in customer friendly voice? I mean what it does?

File: app/controllers/messages_controller.rb (hunk 2 above)
Comment (leonardk1) [inline app/controllers/messages_controller.rb]: @godfreym what happens here is that this method, simply get the participant and then saves their answer depending on their previous reply.

and if there are some more questions to be answered, they are sent to the participant again

File: app/models/participant.rb (hunk 3 above)
Comment (leonardk1) [inline app/models/participant.rb]: @godfreym this instance method here, simply check if the current participant has completed answering all the questions in the campaign

depending on the result, then we do some other processing. in other methods

File: app/controllers/messages_controller.rb (hunk 2 above)
Comment (godfreym) [inline app/controllers/messages_controller.rb]: @leonardk1 OK

File: app/models/participant.rb (hunk 3 above)
Comment (godfreym) [inline app/models/participant.rb]: Cool

**Comment by leonardk1** [comment]
@arafatm please review

**Comment by arafatm** [comment]
:+1:

**Comment by arafatm** [comment]
@leonardk1 will you port this to furaha/ ?

**Pull request merged**